storage/remote/queue_manager.go

```diff
@@ -800,6 +806,35 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue chan sample) {
 			s.droppedOnHardShutdown.Add(uint32(droppedSamples))
 			return
 
+		case <-s.reshard:
+			for {
+				if s.qm.blockIncomingSamples.Load() {
+					// The incoming samples will now be blocked for enty into the new shards,
+					// so its sample to stop listening samples.
+					break
+				}
+				select {
+				case sample, ok := <-queue:
+					// Accept those samples that might have entered the queue between sending the resharding
+					// signal and blocking of incoming samples at the time of creation of newer queues.
+					if !ok {
+						break
+					}
+					pendingSamples[nPending].Labels = labelsToLabelsProto(sample.labels, pendingSamples[nPending].Labels)
+					pendingSamples[nPending].Samples[0].Timestamp = sample.t
+					pendingSamples[nPending].Samples[0].Value = sample.v
+					nPending++
+				default:
+				}
+			}
+			if nPending > 0 {
+				level.Info(s.qm.logger).Log("msg", fmt.Sprintf("Sending %d samples to newer queues.", nPending))
+				s.qm.metrics.pendingSamples.Add(float64(nPending))
+				s.queueBuf[shardID] <- getSamples(nPending, pendingSamples)
+			}
+			close(s.queueBuf[shardID])
```

Comment:
In this path where is `s.queue` closed? It doesn't look like `s.stop()` is ever called for the old shards on a reshard?

Reply:
Yes, I think I left that. `s.queue` should definitely closed.